File: python/Testing/test_mapquest.py
# Importing modules
import json
import requests
import polyline as poly
import os

MAPQUEST_API_KEY = os.environ.get("MAPQUEST_API_KEY")
MAPQUEST_API_URL = "http://www.mapquestapi.com/directions/v2/route"

TOLLGURU_API_KEY = os.environ.get("TOLLGURU_API_KEY")
TOLLGURU_API_URL = "https://apis.tollguru.com/toll/v2"
POLYLINE_ENDPOINT = "complete-polyline-from-mapping-service"

# Explore https://tollguru.com/toll-api-docs to get best of all the parameter that TollGuru has to offer
request_parameters = {
    "vehicle": {
        "type": "2AxlesAuto"
    },
    # Visit https://en.wikipedia.org/wiki/Unix_time to know the time format
    "departure_time": "2021-01-05T09:46:08Z",
}

# Fetching Polyline from mapquest
def get_polyline_from_mapquest(source, destination):
    url = "{a}?key={b}&from={c}&to={d}&fullShape=true".format(
        a=MAPQUEST_API_URL,
        b=MAPQUEST_API_KEY,
        c=source,
        d=destination,
    )
    # Converting the response to JSON
    response = requests.get(url).json()
    # Extracting all the coordinates and making lat-lon pair
    coordinate_list = []
    for i in range(0, len(response["route"]["shape"]["shapePoints"]), 2):
        coordinate_list.append(
            (
                response["route"]["shape"]["shapePoints"][i],
                response["route"]["shape"]["shapePoints"][i + 1],
            )
        )
    # We will encode these coordinates (lat-lon) using encode function from polyline module to generate polyline
    polyline_from_mapquest = poly.encode(coordinate_list)
    return polyline_from_mapquest


# Calling TollGuru API
def get_rates_from_tollguru(polyline):
    # TollGuru query URL
    Tolls_URL = f"{TOLLGURU_API_URL}/{POLYLINE_ENDPOINT}"
    # TollGuru request parameters
    headers = {"Content-type": "application/json", "x-api-key": TOLLGURU_API_KEY}
    params = {
        # Explore https://tollguru.com/developers/docs/ to get best of all the parameter that TollGuru has to offer
        "source": "mapquest",
        "polyline": polyline,  # This is the encoded polyline that we made
        **request_parameters,
    }
    # Requesting TollGuru with parameters
    response_tollguru = requests.post(
        Tolls_URL, json=params, headers=headers, timeout=200
    ).json()
    # print(response_tollguru)
    # checking for errors or printing rates
    if str(response_tollguru).find("message") == -1:
        return response_tollguru["route"]["costs"]
    else:
        raise Exception(response_tollguru["message"])


"""Testing"""
# Importing Functions
from csv import reader, writer
import time

temp_list = []
with open("testCases.csv", "r") as f:
    csv_reader = reader(f)
    for count, i in enumerate(csv_reader):
        # if count>2:
        #   break
        if count == 0:
            i.extend(
                (
                    "Input_polyline",
                    "Tollguru_Tag_Cost",
                    "Tollguru_Cash_Cost",
                    "Tollguru_QueryTime_In_Sec",
                )
            )
        else:
            try:
                polyline = get_polyline_from_mapquest(i[1], i[2])
                i.append(polyline)
            except:
                i.append("Routing Error")

            start = time.time()
            try:
                rates = get_rates_from_tollguru(polyline)
            except:
                i.append(False)
            time_taken = time.time() - start
            if rates == {}:
                i.append((None, None))
            else:
                try:
                    tag = rates["tag"]
                except:
                    tag = None
                try:
                    cash = rates["cash"]
                except:
                    cash = None
                i.extend((tag, cash))
            i.append(time_taken)
        # print(f"{len(i)}   {i}\n")
        temp_list.append(i)

with open("testCases_result.csv", "w") as f:
    writer(f).writerows(temp_list)

"""Testing Ends"""

File: PHP/test/test_mapquest.php
<?php
//using mapquestmaps API

$MAPQUEST_API_KEY = getenv('MAPQUEST_API_KEY');
$MAPQUEST_API_URL = "http://www.mapquestapi.com/directions/v2/route";

$TOLLGURU_API_KEY = getenv('TOLLGURU_API_KEY');
$TOLLGURU_API_URL = "https://apis.tollguru.com/toll/v2";
$POLYLINE_ENDPOINT = "complete-polyline-from-mapping-service";

// Explore https://tollguru.com/toll-api-docs to get the best of all the parameters that Tollguru has to offer
$request_parameters = array(
  "vehicle" => array(
      "type" => "2AxlesAuto"
  ),
  // Visit https://en.wikipedia.org/wiki/Unix_time to know the time format
  "departure_time" => "2021-01-05T09:46:08Z"
);

// From and To locations
function getPolyline($from, $to){
  global $MAPQUEST_API_KEY, $MAPQUEST_API_URL;

  $url=$MAPQUEST_API_URL.'?key='.$MAPQUEST_API_KEY.'&from='.urlencode($from).'&to='.urlencode($to).'&fullShape=true';
  //connection..
  $mapquest = curl_init();

  curl_setopt($mapquest, CURLOPT_SSL_VERIFYHOST, false);
  curl_setopt($mapquest, CURLOPT_SSL_VERIFYPEER, false);

  curl_setopt($mapquest, CURLOPT_URL, $url);
  curl_setopt($mapquest, CURLOPT_RETURNTRANSFER, true);

  // Getting response from MapQuest API
  $response = curl_exec($mapquest);
  $err = curl_error($mapquest);

  curl_close($mapquest);

  if ($err) {
    echo "cURL Error #:" . $err;
  } else {
    echo "200 : OK\n";
  }
  // Extracting polyline from the JSON response
  $data_mapquest = json_decode($response, true);
  $shape_points=$data_mapquest['route']['shape']['shapePoints'];

  // Polyline
  require_once(__DIR__.'/Polyline.php');
  $p_mapquest = Polyline::encode($shape_points);

  return $p_mapquest;
}

// Calling getPolyline function
// Testing starts here
require_once(__DIR__.'/test_location.php');
foreach ($locdata as $item) {
$polyline_mapquest = getPolyline($item['from'], $item['to']);

// Using TollGuru API..
$curl = curl_init();

curl_setopt($curl, CURLOPT_SSL_VERIFYHOST, false);
curl_setopt($curl, CURLOPT_SSL_VERIFYPEER, false);

$postdata = array(
	"source" => "mapquest",
	"polyline" => $polyline_mapquest,
  ...$request_parameters
);

// JSON encoding source and polyline to send as postfields
$encode_postData = json_encode($postdata);

curl_setopt_array($curl, array(
  CURLOPT_URL => $TOLLGURU_API_URL . "/" . $POLYLINE_ENDPOINT,
  CURLOPT_RETURNTRANSFER => true,
  CURLOPT_ENCODING => "",
  CURLOPT_MAXREDIRS => 10,
  CURLOPT_TIMEOUT => 30,
  CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
  CURLOPT_CUSTOMREQUEST => "POST",

  // Sending MapQuest polyline to TollGuru
  CURLOPT_POSTFIELDS => $encode_postData,
  CURLOPT_HTTPHEADER => array(
    "content-type: application/json",
    "x-api-key: " . $TOLLGURU_API_KEY),
));

$response = curl_exec($curl);
$err = curl_error($curl);
curl_close($curl);

if ($err) {
	  echo "cURL Error #:" . $err;
} else {
	  echo "200 : OK\n";
}

// Response from TollGuru
$data = json_decode($response, true);

$tag = $data['route']['costs']['tag'];
$cash = $data['route']['costs']['cash'];

$dumpFile = fopen("dump.txt", "a") or die("unable to open file!");
fwrite($dumpFile, "from =>");
fwrite($dumpFile, $item['from'].PHP_EOL);
fwrite($dumpFile, "to =>");
fwrite($dumpFile, $item['to'].PHP_EOL);
fwrite($dumpFile, "polyline =>".PHP_EOL);
fwrite($dumpFile, $polyline_mapquest.PHP_EOL);
fwrite($dumpFile, "tag =>");
fwrite($dumpFile, $tag.PHP_EOL);
fwrite($dumpFile, "cash =>");
fwrite($dumpFile, $cash.PHP_EOL);
fwrite($dumpFile, "*************************************************************************".PHP_EOL);
echo 'from: '.$item['from'].' to '.$item['to'].'';
echo "\n";
echo "tag = ";
print_r($data['route']['costs']['tag']);
echo "\ncash = ";
print_r($data['route']['costs']['cash']);
echo "\n";
echo "**************************************************************************\n";
}
?>
